pageserver/src/tenant/timeline.rs

```diff
@@ -4176,6 +4177,398 @@ impl Timeline {
+                        &adopted,
+                        &timeline,
+                        timeline.generation,
+                        &timeline.cancel,
+                    )
+                    .await?;
+                    tracing::info!(layer=%owned, "remote copied");
+                    Ok(owned)
+                }
+                .in_current_span(),
+            );
+        }
+
+        while let Some(res) = tasks.join_next().await {
+            match res {
+                Ok(Ok(owned)) => {
+                    new_layers.push(owned);
+                }
+                Ok(Err(failed)) => {
+                    return Err(failed);
+                }
+                Err(je) => return Err(Unexpected(je.into())),
+            }
+        }
+
+        // TODO: fsync directory again if we hardlinked something
+
+        Ok(PreparedTimelineDetach {
+            progress_guard: _guard,
+            layers: new_layers,
+        })
+    }
+
+    /// Completes the ancestor detach. This method is to be called while holding the
+    /// TenantManager's tenant slot, so during this method we cannot be deleted nor can any
+    /// timeline be deleted. After this method returns successfully, tenant must be reloaded.
+    ///
+    /// Pageserver receiving a SIGKILL during this operation is not supported (yet).
+    pub(crate) async fn complete_detaching_timeline_ancestor(
+        self: &Arc<Timeline>,
+        tenant: &crate::tenant::Tenant,
+        prepared: PreparedTimelineDetach,
+        _ctx: &RequestContext,
+    ) -> Result<Vec<TimelineId>, anyhow::Error> {
+        let rtc = self
+            .remote_client
+            .as_ref()
+            .expect("has to have a remote timeline client for timeline ancestor detach");
+
+        let PreparedTimelineDetach {
+            progress_guard: _guard,
+            layers,
+        } = prepared;
+
+        // publish the prepared layers before we reparent any of the timelines, so that on restart
+        // reparented timelines find layers.
+        rtc.schedule_adding_existing_layers_to_index_upload_and_wait(&layers)
+            .await?;
+
+        let mut tasks = tokio::task::JoinSet::new();
+
+        let ancestor = self
+            .get_ancestor_timeline()
+            .expect("must still have a ancestor");
+        let ancestor_lsn = self.get_ancestor_lsn();
+
+        let new_parent = self.timeline_id;
+
+        // because we are now keeping the slot in progress, it is unlikely that there will be any
+        // timeline deletions during this time. if we raced one, then we'll just ignore it.
+        tenant
+            .timelines
+            .lock()
+            .unwrap()
+            .values()
+            .filter_map(|tl| {
+                if Arc::ptr_eq(tl, self) {
+                    return None;
+                }
+
+                if !tl.is_active() {
+                    return None;
+                }
+
+                let tl_ancestor = tl.get_ancestor_timeline().ok()?;
```

Comment:
I have rewritten this a few times, but yeah, risking a useless stacktrace or arc bump makes sense to avoid. Aah yes, for a short period of time this code lived in the manager.

38c856b6a